parser.py

```diff
@@ -166,129 +255,142 @@ class QASMToIRTransformer(Transformer):
     ln = lambda self, _: "log"
     sqrt = lambda self, _: "sqrt"
 
+    def __init__(self, *args, **kwargs):
+        self._program = QasmProgram()
+        super().__init__(self, *args, **kwargs)
+
     def mainprogram(self, *args):
         args = unpack(args)
 
-        if len(args) == 1:
-            program_str = args[0]
-            header = ""
-        else:
-            version, program_str = args
-            header = "OPENQASM {};".format(version)
+        if len(args) > 1:
+            version = args[0]
+            self._program.version = version
 
-        return "\n".join([header, program_str])
+        return self._program
 
     def program(self, *args):
-        args = flatten_recursive_list(*args)
-        program_str = "\n".join(str(x) for x in args)
-        # TODO: convert this function to return data structures as output, not strings
-        return program_str
+        return None
+        # args = flatten_recursive_list(*args)
+        # program_str = "\n".join(str(x) for x in args)
+        # # TODO: convert this function to return data structures as output, not strings
+        # return program_str
```

Comment:
Since the tree is traversed from leaf to root, I create the `QasmProgram` at `__init__` as a tree attribute, and populate it during the traversal. Hence why this node doesn't return anything (can probably be deleted)